tests/main/test_tool_messages.py

```diff
@@ -1187,6 +1197,410 @@ def agent_response(self, msg: str | ChatDocument) -> Any:
+                metadata=ChatDocMetaData(sender=Entity.LLM),
+                oai_tool_calls=[
+                    OpenAIToolCall(
+                        id="call-1234657",
+                        function=attempt,
+                    )
+                ],
+            )
+
+        return ChatDocument(
+            content="",
+            metadata=ChatDocMetaData(sender=Entity.LLM),
+            function_call=attempt,
+        )
+
+    assert (
+        simulate_failed_call(
+            to_attempt(
+                LLMFunctionCall(
+                    name="__nabroski__",
+                    arguments={
+                        "xval": 1,
+                        "yval": 3,
+                    },
+                )
+            )
+        )
+        == "6"
+    )
+    assert (
+        simulate_failed_call(
+            to_attempt(
+                LLMFunctionCall(
+                    name="Nabroski-function",
+                    arguments={
+                        "xval": 2,
+                        "yval": 3,
+                    },
+                )
+            )
+        )
+        == "9"
+    )
+    # Strict fallback disables the default arguments, but the LLM should infer
+    # from context
+    assert (
+        simulate_failed_call(
+            """
+        request ":coriolis"
+        arguments {"xval": 1}
+        """
+        )
+        == "8"
+    )
+    assert (
+        simulate_failed_call(
+            to_attempt(
+                LLMFunctionCall(
+                    name="Coriolis",
+                    arguments={
+                        "xval": 1,
+                    },
+                )
+            )
+        )
+        == "8"
+    )
+    assert (
+        simulate_failed_call(
+            to_attempt(
+                LLMFunctionCall(
+                    name="EulerTool",
+                    arguments={
+                        "xval": 6,
+                        "yval": 4,
+                    },
+                )
+            )
+        )
+        == "8"
+    )
+
+
+@pytest.mark.parametrize("use_fn_api", [True, False])
+@pytest.mark.parametrize("use_tools_api", [True, False])
+@pytest.mark.parametrize("parallel_tool_calls", [True, False])
+def test_strict_fallback(
+    test_settings: Settings,
+    use_fn_api: bool,
+    use_tools_api: bool,
+    parallel_tool_calls: bool,
+):
+    """
+    Test that strict tool and structured output errors
+    are handled gracefully and are disabled if errors
+    are caused.
+    """
+    set_global(test_settings)
+
+    class BrokenStrictSchemaAgent(ChatAgent):
+        def _function_args(self) -> tuple[
+            Optional[List[LLMFunctionSpec]],
+            str | dict[str, str],
+            Optional[list[OpenAIToolSpec]],
+            Optional[dict[str, dict[str, str] | str]],
+            Optional[OpenAIJsonSchemaSpec],
+        ]:
+            functions, fun_call, tools, force_tool, output_format = (
+                super()._function_args()
```

Comment:
If I understood correctly, this is implementing a "broken" version of the correct `_function_args()` -- instead of this, I wonder if there's a more direct way to do the test, e.g. manually set various values?

Reply:
Or maybe add comments to make clear exactly what is broken in this implementation

Reply:
Hmm maybe; we could modify `ChatAgent` to have a configurable schema editing function if that's better. Added a comment.